src/adapter/src/coord/timestamp_oracle/catalog_oracle.rs

```diff
@@ -0,0 +1,294 @@
+pub(crate) struct CatalogTimestampPersistence {
+    timeline: Timeline,
+    catalog: Arc<Catalog>,
+}
+
+impl CatalogTimestampPersistence {
+    pub(crate) fn new(timeline: Timeline, catalog: Arc<Catalog>) -> Self {
+        Self { timeline, catalog }
+    }
+}
+
+#[async_trait::async_trait]
+impl TimestampPersistence<mz_repr::Timestamp> for CatalogTimestampPersistence {
+    async fn persist_timestamp(&self, timestamp: mz_repr::Timestamp) -> Result<(), Error> {
+        self.catalog
+            .persist_timestamp(&self.timeline, timestamp)
+            .await
+    }
+}
+
+/// Convenience function for calculating the current upper bound that we want to
+/// prevent the global timestamp from exceeding.
+// TODO(aljoscha): These internal details of the oracle are leaking through to
+// multiple places in the coordinator.
+pub(crate) fn upper_bound(now: &mz_repr::Timestamp) -> mz_repr::Timestamp {
+    now.saturating_add(
+        TIMESTAMP_PERSIST_INTERVAL.saturating_mul(Timestamp::from(TIMESTAMP_INTERVAL_UPPER_BOUND)),
+    )
+}
+
+/// Returns the current system time while protecting against backwards time
+/// jumps.
+///
+/// The caller is responsible for providing the previously recorded system time
+/// via the `previous_now` parameter.
+///
+/// If `previous_now` is more than `TIMESTAMP_INTERVAL_UPPER_BOUND *
+/// TIMESTAMP_PERSIST_INTERVAL` milliseconds ahead of the current system time
+/// (i.e., due to a backwards time jump), this function will block until the
+/// system time advances.
+///
+/// The returned time is guaranteed to be greater than or equal to
+/// `previous_now`.
+// TODO(aljoscha): These internal details of the oracle are leaking through to
+// multiple places in the coordinator.
+pub(crate) fn monotonic_now(now: NowFn, previous_now: mz_repr::Timestamp) -> mz_repr::Timestamp {
+    let mut now_ts = now();
+    let monotonic_now = cmp::max(previous_now, now_ts.into());
+    let mut upper_bound = catalog_oracle::upper_bound(&mz_repr::Timestamp::from(now_ts));
+    while monotonic_now > upper_bound {
+        // Cap retry time to 1s. In cases where the system clock has retreated
+        // by some large amount of time, this prevents against then waiting for
+        // that large amount of time in case the system clock then advances back
+        // to near what it was.
+        let remaining_ms = cmp::min(monotonic_now.saturating_sub(upper_bound), 1_000.into());
+        error!(
+            "Coordinator tried to start with initial timestamp of \
+            {monotonic_now}, which is more than \
+            {TIMESTAMP_INTERVAL_UPPER_BOUND} intervals of size {} larger than \
+            now, {now_ts}. Sleeping for {remaining_ms} ms.",
+            *TIMESTAMP_PERSIST_INTERVAL
+        );
+        thread::sleep(Duration::from_millis(remaining_ms.into()));
+        now_ts = now();
+        upper_bound = catalog_oracle::upper_bound(&mz_repr::Timestamp::from(now_ts));
+    }
+    monotonic_now
+}
```

Comment:
This function only really makes sense when used with the `EpochMillis` timeline. That timeline isn't mentioned anywhere in this module and feels like it's an implementation detail of other parts of the adapter. For those reasons it feels a little out of place in this module.

Reply:
yeah, I also didn't know what to do about this one. In the end I moved it here with the oracle code because it uses internals of the oracle like `upper_bound` and the `TIMESTAMP_PERSIST_INTERVAL` and `TIMESTAMP_INTERVAL_UPPER_BOUND`. Plus, once we remove this oracle we have more of the code that we need to remove localized in this one file.

But I'm happy to move that back to `timeline.rs`. And I probably have to add more code here that does a `monotonic_now` for both oracles, in the interim period where we have both oracles/where we switch over. What do you think?